Logos and OS information

sjg20 opened the pull request:
Introduce support for logos for each configuration, to allow for a more attractive boot menu.

Also provide OS information, based on the os-release specification, to allow for richer information in the boot menu.

Fixes: #25

sjg20 committed:
Support os-release information

It is useful to have some information about an OS, e.g. to present to
the user in a menu.

Create a definition for this, making use of the os-release
specification. Add a note into the usage section about this.

Co-developed-by: Claude <noreply@anthropic.com>
Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
Add support for logos

Provide a way to specify one or more logos so that the bootloader can
present a nicer graphical menu for the OS.

Support BMP as a baseline, with PNG and SVG as optional features.

Add a note into the usage section about boot menus, with some
recommendations for how a logo should be selected from the available
options.

Signed-off-by: Simon Glass <sjg@chromium.org>

xypron commented on source/chapter2-source-file-format.rst:
```diff
@@ -514,6 +526,404 @@ padding
+The 'logo-info' node has the following structure::
+
+    o logo-info
+       |- format = "bmp";
```
Please, do not define a structure via an example but by field names and value types and value ranges.
The Linux documentation for device-trees shows how to do this properly.

I would suggest to refer to https://www.iana.org/assignments/media-types/media-types.xhtml.

You seem to suggest to use the media sub-type (e.g. bmp) of an image type instead of the full mime-type (e.g. image/bmp). This would exclude using a video as a logo in future.

sjg20 replied:
done

xypron commented on source/chapter2-source-file-format.rst:
```diff
@@ -514,6 +526,404 @@ padding
+    bmp             Mandatory  Windows bitmap; see [windowsbmp]_
+    png             Optional   Portable Network Graphics; see [png]_
+    svg             Optional   Scalable Vector Graphics; see [svg]_
+    ==============  =========  =====================================
```
svg images may include ECMA script code. Do expect to support that in firmware?

I would suggest to look for the file formats supported in ACPI BGRT tables.

sjg20 replied:
OK let's just drop it for now

xypron commented on source/chapter2-source-file-format.rst:
```diff
@@ -514,6 +526,404 @@ padding
+depth-log2
+    log (base 2) of the logo depth. For example, for a 32bpp images this should
+    be 5. This property is mandatory for logos which define a colour depth,
+    i.e. BMP and PNG.
```
The depth is encoded in the file. There is not need to put it here.

BMP files allow 15 bit depth 5,5,5. You can have RGB or BGR encoding.

sjg20 replied:
removed

xypron commented on source/chapter2-source-file-format.rst:
```diff
@@ -514,6 +526,404 @@ padding
+    indexed         A palette is used to define colours
+    grayscale       The image supports luminance only
+    rgb             The image supports red/green/blue true colour
+    ==============  =============================================
```
Why do we need this information here? It is encoded in the image and only needed by the display code.

sjg20 replied:
removed

xypron commented on source/chapter2-source-file-format.rst:
```diff
@@ -514,6 +526,404 @@ padding
+The 'os-info' node has the following structure::
+
+    o os-info
+        |- ansi-color = "OS console presentation color"
```
The kernel log is always shown on black background. The background color after login is user customizable. What is the use of this field?

sjg20 replied:
removed

xypron commented on source/chapter2-source-file-format.rst:
```diff
@@ -514,6 +526,404 @@ padding
+~~~~~~~~~~~~~~~~~~~
+
+ansi-color
+    A suggested presentation color when showing the OS name on the console.
```
Why wouldn't a firmware use its preferred color?
Do you want to use different front and backgrounds for a list of OSes in a menu?

xypron commented on source/chapter2-source-file-format.rst:
```diff
@@ -514,6 +526,404 @@ padding
+    A lower-case string identifying the operating system configuration
+    extensions support level, used to specify extension-release
+    compatibility.
+
```
It is unclear what information you want to convey here.
What should a firmware do with it?

xypron commented on source/chapter2-source-file-format.rst:
```diff
@@ -514,6 +526,404 @@ padding
+cpe-name
+    A CPE name for the operating system, in URI binding syntax, following
+    the Common Platform Enumeration Specification.
+
```
Please, provide a link to the specification.

xypron commented on source/chapter2-source-file-format.rst:
```diff
@@ -514,6 +526,404 @@ padding
+
+    Example::
+        default-hostname = "localhost";
+
```
What do you want to do with this? 
Why should a firmware care?

xypron commented on source/chapter2-source-file-format.rst:
```diff
@@ -514,6 +526,404 @@ padding
+    ".", "_" and "-".
+
+    Examples::
+        id = "fedora";
```
Isn't this redundant to cpe-name?

xypron commented on source/chapter2-source-file-format.rst:
```diff
@@ -514,6 +526,404 @@ padding
+    operating system in regards to packaging and programming interfaces.
+
+    Examples::
+        id-like = "rhel", "centos", "fedora";
```
What is the use of this field? It seems straight out superfluous.

xypron commented on source/chapter2-source-file-format.rst:
```diff
@@ -514,6 +526,404 @@ padding
+
+    Examples::
+        logo-name = "fedora-logo-icon";
+
```
The firmware typically will not be able to access the root file system.
This field is superfluous.

xypron commented on source/chapter2-source-file-format.rst:
```diff
@@ -514,6 +526,404 @@ padding
+        pretty-name = "Fedora Linux 32 (Workstation Edition)";
+
+        pretty-name = "Ubuntu 20.04.1 LTS";
+
```
Please, don't introduce dozens of redundant fields.

sjg20 replied:
Updated to the bare minimum needed to show the menu

xypron commented on source/chapter2-source-file-format.rst:
```diff
@@ -514,6 +526,404 @@ padding
+    one.
+
+    Example::
+        privacy-policy-url = "https://fedoraproject.org/wiki/Legal:PrivacyPolicy";
```
Or do you want to show a QR code in firmware to let the user read the webpage?

sjg20 replied:
Yes, good idea, added mention of that

xypron commented on source/chapter2-source-file-format.rst:
```diff
@@ -514,6 +526,404 @@ padding
+        release-type = "lts";
+
+        release-type = "development";
+
```
How would you encode rolling release operating systems like Arch?

sjg20 replied:
removed for now...this is something in the osrelease spec but I don't think we need it

xypron commented on source/chapter2-source-file-format.rst:
```diff
@@ -514,6 +526,404 @@ padding
+
+    Example::
+        support-end = /bits 64/ <1756212516>;
+
```
The creation time is of much more interest for rolling releases.

xypron replied:
Which support-end would you enter here for Ubuntu:
* end of free support
* end of paid support via Ubuntu Pro

Debian has even three end of life dates for each release in https://wiki.debian.org/DebianReleases

As the field has no clear definition it should be removed.

sjg20 replied:
removed for now...we can always revisit if we need it

xypron commented on source/chapter2-source-file-format.rst:
```diff
@@ -514,6 +526,404 @@ padding
+sysext-level
+    A lower-case string identifying the operating system extensions support
+    level, used to specify extension-release compatibility.
+
```
Are you relating to any specification?

sjg20 replied:
removed for now

sjg20 committed:
Address review feedback on logos and os-info

Address review feedback from xypron on PR #29.

Logo-info node:
- Use IANA media types (e.g. "image/bmp") instead of short names
- Remove redundant fields (width, height, depth-log2, color) that are
  already encoded in the image file
- Remove SVG support due to ECMA script concerns in firmware
- Keep only format (mandatory) and background (optional)

OS-info node:
- Remove fields that firmware cannot use or that are superfluous:
  architecture, bug-report-url, build-id, confext-level, cpe-name,
  default-hostname, documentation-url, logo-name, privacy-policy-url,
  support-end, support-url, sysext-level, ansi-color, color,
  release-type
- Rename 'variant' to 'os-variant' to avoid conflict with image variant
  property
- Keep fields useful for boot menus: name, version, pretty-name, id,
  id-like, image-id, image-version, home-url, os-variant, variant-id,
  vendor-name, vendor-url, version-codename, version-id
- Note that URLs may be displayed as QR codes
- Fix RST formatting (blank lines before examples)

References:
- Add IANA media types reference
- Remove unused SVG and icontheme references

Co-developed-by: Claude Opus 4.5 <noreply@anthropic.com>
Signed-off-by: Simon Glass <simon.glass@canonical.com>

orangecms commented on source/chapter2-source-file-format.rst:
```diff
@@ -393,6 +396,15 @@ signature-1
 
+logo-info
+    Contains information about the logo image, to help describe characteristics
+    such as image dimensions.
```
Remove; below it says the dimensions are taken from parsing the image.

Can we just called it `logo` instead of `logo-info`? The "info" part suggests extra information, but the image formats specified here already carry their own metadata.

orangecms commented on source/chapter3-usage.rst:
```diff
@@ -134,6 +134,39 @@ found, searching stops, using the best match found in the stage.
+
+    #. The logo format must be supported by the bootloader
+
+    #. SVG is preferred (if supported), followed by PNG (also if supported)
```
SVG was removed in an iteration here due to the question re ECMAScript, drop it here as well.

orangecms reviewed (commented):
I like this, would love to see it move forward.

Added a few remarks for consistency, based on briefly skimming.

There is a little conflict now, so a rebase is necessary.

Thank you! 🧡
